### PR TITLE
persist: avoid decoding persist data when data are ignored

### PR DESCRIPTION
**Pull request by danhhz**
Reviving #25994

Avoid decoding data when the `MfpPlan` tells us we will not read the results. This is a total hack, pending columnar data representations, but at the moment it reduces `main` from (release, local not cloud)
```
materialize=> select count(*) from bar;
  count   
----------
 10000000
(1 row)

Time: 2719.868 ms (00:02.720)
materialize=> 
```
to (in this branch)
```
materialize=> select count(*) from bar;
  count   
----------
 10000000
(1 row)

Time: 1153.915 ms (00:01.154)
materialize=> 
```

Touches #20618

### Motivation

  * This PR adds a feature that has not yet been specified.

### Tips for reviewer

The first commit is exactly Frank's original PR, except rebased.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-operators/src/persist_source.rs
Comment (danhhz) [inline src/storage-operators/src/persist_source.rs:634]: whoops, "or None if it's unknown"

**Comment by danhhz** [comment]
Okay @bkirwi I'd like to start merging the `select count(*) from persist` work, so time to give this one a review, please! I'll add a CYA dyncfg to this before merging, but holding off on that until the first round of code review comments are in

File: src/storage-operators/src/persist_source.rs
Comment (bkirwi) [inline src/storage-operators/src/persist_source.rs:476]: Good naming!

File: src/persist-client/src/fetch.rs
Comment (bkirwi) [inline src/persist-client/src/fetch.rs:648]: A bit of a surprising API, but all the alternatives I can imagine are much more invasive, so this feels like the right call to me also.

**Review by bkirwi — APPROVED** [review]
Looks good!

**Comment by danhhz** [comment]
TFTR! Pushed the CYA dyncfg and going to run a nightlies before merge

**Pull request merged**
